js/src/product-feed/issues-table-card/issues-table.js

```diff
@@ -0,0 +1,97 @@
+/**
+ * External dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { CardBody, CardFooter } from '@wordpress/components';
+import { Pagination, TablePlaceholder } from '@woocommerce/components';
+
+/**
+ * Internal dependencies
+ */
+import { ISSUE_TABLE_PER_PAGE } from '../constants';
+import { recordTablePageEvent } from '.~/utils/recordEvent';
+import useMCIssues from '.~/hooks/useMCIssues';
+import getActiveIssueType from '.~/product-feed/issues-table-card/getActiveIssueType';
+import IssuesTableData from '.~/product-feed/issues-table-card/issues-table-data';
+import './index.scss';
+
+const headers = [
+	{
+		key: 'type',
+		label: __( 'Type', 'google-listings-and-ads' ),
+		isLeftAligned: true,
+		required: true,
+	},
+	{
+		key: 'affectedProduct',
+		label: __( 'Affected product', 'google-listings-and-ads' ),
+		isLeftAligned: true,
+		required: true,
+	},
+	{
+		key: 'issue',
+		label: __( 'Issue', 'google-listings-and-ads' ),
+		isLeftAligned: true,
+		required: true,
+	},
+	{
+		key: 'suggestedAction',
+		label: __( 'Suggested action', 'google-listings-and-ads' ),
+		isLeftAligned: true,
+		required: true,
+	},
+	{ key: 'action', label: '', required: true },
+];
+
+/**
+ * The table rendering the issues data with a paginator.
+ * It uses useMCIssues for filtering the issues
+ * It uses getActiveIssueType in order to know which issue is active.
+ *
+ * @see useMCIssues
+ * @see getActiveIssueType
+ * @return {JSX.Element} The rendered component
+ */
+const IssuesTable = () => {
+	const issueTypes = useMCIssues();
+	const issues = issueTypes[ getActiveIssueType() ];
```

Comment:
Using useMCIssuesTypeFilter directly would override the paginator state on each switch. I would continue with the current system.

Reply:
> Using `useMCIssuesTypeFilter` directly would override the paginator state on each switch.

That was my concern. The `page` and `setPage` returned from `useMCIssuesTypeFilter` hook were not corresponding to the current passing `issueType` parameter. It makes the use of `useMCIssuesTypeFilter` subject to this implied restriction, and this restriction does not match the functional needs of the scenario.

Since `useMCIssuesTypeFilter` is a standalone hook, I think it would be better to externalize `page` and `setPage` to consumer sides, or associate them with `issueType`. Or keep the current implementation and internalize `useMCIssuesTypeFilter` into `useMCIssues`.

Reply:
💅  Sorry. Not sure if I understand the idea here

Reply:
Not sure which parts are unclear?

Considering the `useMCIssuesTypeFilter` is a standalone hook, it's expected that it can be used directly without using it through the `useMCIssues` always. Currently, it provides `issueType` argument, however, the state of the returned `page` is shared between issue types. The actual data sources of the issue types are different collections, and their page states should be independent of each other.

> I think it would be better to externalize `page` and `setPage` to consumer sides, or associate them with `issueType`. Or keep the current implementation and internalize `useMCIssuesTypeFilter` into `useMCIssues`.

Even though the runtime results of current implementation are correct, the open interface of `useMCIssuesTypeFilter` hides this problem. The above approaches were proposed in order to avoid this problem when using `useMCIssuesTypeFilter` directly in the future.

The last suggestion is to move `useMCIssuesTypeFilter` into `useMCIssues` and use it as an internal hook in the module, and don't export it. Although this isn't the most suitable design solution in my opinion, it's probably a relatively convenient choice at the moment. After all, the two data sources of issue types are actually always fetched together.

Reply:
So, the idea of `useMCIssuesTypeFilter` as a standalone hook is for the case you want to use it as standalone hook. 

1. Example use case (real): 

When you click in the future "Review account" button, It will show a modal with the list of account issues (only account issues, not Product issues) 

2. Example use case (not real): 

If we would add a new Issues nav section at the top level. That is, separating Product feed from Issues. Then we would show maybe 2 tables in this case, 1 for account and 1 for product. Without the Switch. That way yes, you can use as well `useMCIssuesTypeFilter` as standalone hook.

The idea behind `useMCIssues` (without params now) is for grouping  the `useMCIssuesTypeFilter` hooks and also for adding the totals (recently). This is necessary because each issueType needs to have their own internal paginator persistent after switches. So, actually it's doing that association you mentioned between the paginator and the issueType, in other words, useMCIssues is creating 2 instances of `useMCIssuesTypeFilter` with their own paginator. But you can always use the standalone version for cases like 1 and 2. 

One solution I had at some point in this PR and was discarded was this:

```
   // index.js 
   <IssuesTable issueType="ISSUE_TYPE_ACCOUNT" />
   <IssuesTable issueType="ISSUE_TYPE_PRODUCT" />
```
```
    // issues-table.js   
    const { data, hasFinishedResolution, page, setPage } = useMCIssuesTypeFilter( issueType )
    if (getActiveIssueType() !== issueType) return null;
    ...
```

But felt for me redundant, and the current solution looked for me cleaner.
 
> Not sure which parts are unclear?

> I think it would be better to externalize page and setPage to consumer sides, or associate them with issueType

Mainly this, in my mind this potential implementation ends having a mapping function for associate each issueType with a different paginator with multiple state management, or kinda the same logic outside the hook. 

But I think it could be because is not clear for me the implementation of the proposed idea.

Reply:
> Using `useMCIssuesTypeFilter` directly would override the paginator state on each switch.

But the `useMCIssuesTypeFilter` couldn't be used standalone for the [original intention](https://github.com/woocommerce/google-listings-and-ads/pull/1176#discussion_r777992907) of this thread, and it's a real use case.

Based on both issues data are also fetched at the same time by usages in other components. Even though changing the `<IssuesTable>` component to use `useMCIssuesTypeFilter` will not reduce any network loading, from component design perspective of reusability, it's not so proper to make it ask for all data sources but actually only needs one of those data sources to complete its functions.

Given an example that a `<AccountInfo>` component is used to display general account information, and the account data have a dozen sources and have to fetch from server APIs independently in each other. When using `<AccountInfo type="jetpack" />`, it triggers a dozen requests to server for getting all the account information and then picks out the jetpack to render. I think this's apparently not a proper component design of reusability in most cases. Even if it needs preloading feature in some scenarios, usually the logic is probably not placed inside that component, but in an upper wrapper or handled by the data layer of somewhat client service.

---

Speaking of `useMCIssues` and `useMCIssuesTypeFilter`, I also think the `useMCIssues` is needed but in a bit different purposes thinking. I would clearly distinguish their different purposes in this way.

- `useMCIssuesTypeFilter` as data hook serves MC issues retrieving oriented purpose.
    - Provides issues data according to arguments of type, page and per page.
    - Doesn't have the internal query state of `page`.
- `useMCIssues` as combined information hook serves calculated states for UI-oriented purpose.
    - Provides such as the total number and data loading state across all issues.
    - And also each total number, or whether at least one data loading is complete and having issues.
    - Doesn't provide query parameters and does not expose issues array.

In this way, it'd be clear which one should be adopted when developing components.

> The idea behind `useMCIssues` (without params now) is for grouping the `useMCIssuesTypeFilter` hooks and also for adding the totals (recently). This is necessary because each issueType needs to have their own internal paginator persistent after switches. So, actually it's doing that association you mentioned between the paginator and the issueType, in other words, useMCIssues is creating 2 instances of `useMCIssuesTypeFilter` with their own paginator.

This works but also makes `useMCIssuesTypeFilter` look unnecessary to be a standalone hook. Because
- no component uses it directly in the present.
- there's a component, which is actually more proper to directly use it, would have a problem, so it couldn't.

---

> Mainly this, in my mind this potential implementation ends having a mapping function for associate each issueType with a different paginator with multiple state management, or kinda the same logic outside the hook.

Yes.

> But I think it could be because is not clear for me the implementation of the proposed idea.

**Associate internal state with `issueType`**

```js
const useMCIssuesTypeFilter = ( issueType = ISSUE_TYPE_ACCOUNT ) => {
	const [ pageHolder, setPageHolder ] = useState( {} );
	const { [ issueType ]: page = defaultQuery.page } = pageHolder;

	const { data, hasFinishedResolution } = useAppSelectDispatch(
		'getMCIssues',
		{
			...defaultQuery,
			page,
			issue_type: issueType,
		}
	);

	const setPage = useCallback(
		( nextPage ) => {
			setPageHolder( ( prevPage ) => ( {
				...prevPage,
				[ issueType ]: nextPage,
			} ) );
		},
		[ issueType ]
	);

	return { data, hasFinishedResolution, page, setPage };
};
```

**Externalize `page` and `setPage` to consumer sides**

```js
// useMCIssuesTypeFilter.js
const useMCIssuesTypeFilter = (
	issueType,
	page = 1,
	perPage = ISSUE_TABLE_PER_PAGE
) => {
	return useAppSelectDispatch( 'getMCIssues', {
		issue_type: issueType,
		page,
		per_page: perPage,
	} );
};

// issues-table.js
const IssuesTable = () => {
	const issueType = getActiveIssueType();
	const [ pageHolder, setPageHolder ] = useState( {} );
	const { [ issueType ]: page = 1 } = pageHolder;
	const issues = useMCIssuesTypeFilter( issueType, page );

	if ( ! issues ) return null;

	const { data, hasFinishedResolution } = issues;

	const handlePageChange = ( newPage, direction ) => {
		setPageHolder( {
			...pageHolder,
			[ issueType ]: newPage,
		} );
		// ... omitted
	};
	// ... omitted
};
```

Both are using the same methods. I'd prefer the externalization one more. The reasons are:

1. Most use cases don't need the `page` state.
2. Further considering, the `page` could be parts of URL query parameters. Given a scenario that may have happened:
    - There're hundreds of product issues in a merchant's store. There're hundreds of product issues in a merchant's store. The owner decides to allocate those issues to staff to address.
    - Since it only shows 5 records per page, the owner takes some time to arrive at the 24th page and passes the URL of the current page to employee A as the start position of divided tasks.
    - Employee A opens the URL and notices it's on the first page.
    - Ok, they realize it doesn't locate to the same viewing page states.
    - Oh no, it's really noisy that it can't jump to the 24th page
    - They have to redo the switching to the 24th page.
    - Or, maybe even Employee A didn't aware it doesn't the right page and started working from the first page.
    - A few hours later, they finally found that their work had overlapped and were very frustrated. 😞 

Surely making the URL sync with page viewing states is a 📜 thought and beyond scope of this PR. But externalization option makes it easier to adjust in case we want to take care of this kind of scenario.

Reply:
Thanks @eason9487. Looks good the second for me as well, thanks for detailing the implementation, now is clear for me. 

Before implementing this, I have a few suggestions/doubts I want to discuss. 

1. `if ( ! issues ) return null;` What about remove it hand just handle the data and the resolution inside `useMCIssuesTypeFilter`
2.  I guess `setPageHolder` logic can be used as  const {page, setPage} = useIssuesPagination( issueType ); Need to test it. 
3. `useMCIssuesTypeFilter ` feels redundant for me now, since is just a Proxy for `useAppSelectDispatch( 'getMCIssues')`
4. I would also change then `mcIssues` as suggested, renaming it to `mcIssuesTotals` and providing only the totals, maybe wrapped by some callback to dont re-calculate the totals since its always same totals right?

Reply:
> * `if ( ! issues ) return null;` What about remove it hand just handle the data and the resolution inside `useMCIssuesTypeFilter`

No matter which solutions, that `issues` is always an object, and that `if` conditional statement is never true. 😂  Looks like the only need is to remove it without other relevant changes.

> * I guess `setPageHolder` logic can be used as  const {page, setPage} = useIssuesPagination( issueType ); Need to test it.

Yes, it could. 👍  And somewhat it looks nicer because it's could be a general hook for wider reusability.

```js
const usePage = ( handle, defaultPage = 1 ) => {
	const [ pageHolder, setPageHolder ] = useState( {} );
	const { [ handle ]: page = defaultPage } = pageHolder;
	const setPage = useCallback(
		( nextPage ) => {
			setPageHolder( ( prevPage ) => ( {
				...prevPage,
				[ handle ]: nextPage,
			} ) );
		},
		[ handle ]
	);
	return [ page, setPage ];
};
```

> * `useMCIssuesTypeFilter ` feels redundant for me now, since is just a Proxy for `useAppSelectDispatch( 'getMCIssues')`

It depends on whether those default values of `page` and `perPage` arguments are useful in your thoughts. AFAIK, another thing is the general `useAppSelectDispatch` hook might not have a work way for indicating the real JSDoc of the selector to be called. So having that hook and arguments JSDoc is still a bit helpful in a manner of developer experience. 🤔 

> * I would also change then `mcIssues` as suggested, renaming it to `mcIssuesTotals` and providing only the totals, maybe wrapped by some callback to dont re-calculate the totals since its always same totals right?

It'd be great in my opinion that It'd help components focus on how to render the delivered states but not about how to calculate the states. Another benefit perhaps it'd be easier to write tests for the aggregation logics rather than testing them via components. About the wrapping callback (`useMemo`?), since it should be no reference changes problem in every rendering on total values. Whether there is a significant benefit from reducing the number of calculations needs to be evaluated through some tools. Or maybe add it when finding the rendering performance is really a concern.

Reply:
Howdy again @eason9487 

I added here https://github.com/woocommerce/google-listings-and-ads/pull/1176/commits/12dce2377e7376abfb186b53d1890adbce12689b a few modifications

- created `usePagination` hook. (It's actually pretty similar to your proposal, but happy to adapt it again if needed)
- adapted `useMCIssuesTypeFilter`
- adapted and renamed `useMCIssues` now `useMCIssuesTotals` 
- adapted the tests

Re - About the wrapping callback . The idea was to avoid is recalcuating the totals each time issues page change.  Reason of this is that totals and issues use the same dispatcher. So on each new page or change in the issues it also triggers the totals subscription.

If there is no more suggestions or improvements in this code, I would ask for a new brief test before merge. 

Thanks 🙏